delphi-checks/src/main/java/au/com/integradev/delphi/utils/FormatStringUtils.java

```diff
@@ -0,0 +1,429 @@
+/*
+ * Sonar Delphi Plugin
+ * Copyright (C) 2023 Integrated Application Development
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package au.com.integradev.delphi.utils;
+
+import java.text.CharacterIterator;
+import java.text.StringCharacterIterator;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public final class FormatStringUtils {
```

Comment:
Let's think about the API a bit more.

I don't love that we've got a bunch of imports that look like:
```java
import au.com.integradev.delphi.utils.FormatStringUtils.DelphiFormatArgument;
import au.com.integradev.delphi.utils.FormatStringUtils.DelphiFormatString;
import au.com.integradev.delphi.utils.FormatStringUtils.FormatSpecifierType;
```
Specifically, I'm not a big fan of `Utils` classes that export types and expect you to drag them out and use them directly. Grouping classes together is what packages are for, and we can just as easily hide the private classes with package-private visibility.

I wonder if it would make more sense to:
- pull all of these nested types out into a `utils.format` package
- make `FormatStringParser::parse` return a straight-up `DelphiFormatString`

I think that's the minimum shape we'd need this in, if we ever wanted to expose format string parsing via the public API in the future.

Reply:
I'm happy to do this, it will be good to establish a precedent for more detailed check utilities. I also don't love utility classes that are really namespaces.

It's worth mentioning, though, that this won't ever be part of the public API, since it's in `delphi-checks`.

Reply:
> this won't ever be part of the public API

Never say never - if developers requested it in the public API, we could easily do it.
Even more easily, if we make these changes (since I'd be happy with the overall shape for a public API)

Let's do it. 👍

Reply:
Very true! Reworked in my latest push.